File: microsynth/microsynth/report/customer_finder/customer_finder.js
// Copyright (c) 2022, Microsynth, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Customer Finder"] = {
    "filters": [
        {
            "fieldname": "contact_name",
            "label": __("Person ID"),
            "fieldtype": "Link",
            "options": "Contact"
        },
        {
            "fieldname": "contact_full_name",
            "label": __("Contact Name"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "contact_email",
            "label": __("Contact Email"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "customer",
            "label": __("Customer (Company/Uni)"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "customer_id",
            "label": __("Customer ID"),
            "fieldtype": "Data",
            "options": "",
            "hidden": 1
        },
        {
            "fieldname": "contact_institute",
            "label": __("Institute"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "contact_institute_key",
            "label": __("Institute Key"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "contact_department",
            "label": __("Department"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "contact_group_leader",
            "label": __("Group Leader"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "address_country",
            "label": __("Country"),
            "fieldtype": "Link",
            "options": "Country"
        },
        {
            "fieldname": "address_city",
            "label": __("City"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "address_street",
            "label": __("Street"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "price_list",
            "label": __("Price List"),
            "fieldtype": "Link",
            "options": "Price List"
        },
        {
            "fieldname": "account_manager",
            "label": __("Sales Manager"),
            "fieldtype": "Link",
            "options": "User"
        },
        {
            "fieldname": "tax_id",
            "label": __("Tax ID"),
            "fieldtype": "Data",
            "options": ""
        },
        {
            "fieldname": "contact_status",
            "label": __("Contact Status"),
            "fieldtype": "Select",
            "options": [ '', 'Passive', 'Open', 'Lead', 'Replied', 'Disabled']
        },
        {
            "fieldname": "contact_classification",
            "label": __("Contact Classification"),
            "fieldtype": "Select",
            "options": [ '', 'Buyer', 'Former Buyer', 'Lead']
        },
        {
            "fieldname": "customer_status",
            "label": __("Customer Status"),
            "fieldtype": "Select",
            "options": [ '', 'active', 'former', 'potential']
        },
        {
            "fieldname":"include_disabled",
            "label": __("Include disabled Customers (e.g. leads)"),
            "fieldtype": "Check"
        }
    ],
    "onload": (report) => {
        hide_chart_buttons();
        var btn = report.page.add_inner_button(__('New Lead'), function () {
            open_new_lead_dialog();
        });
        btn.addClass('btn-success');
    },
    'after_datatable_render': function(report) {
        const applyRowColors = () => {
            const rows = document.querySelectorAll('[data-row-index]');
            rows.forEach(row => {
                const index = parseInt(row.getAttribute('data-row-index'), 10);
                const rowData = frappe.query_report.data[index];
                if (!rowData) return;
                const highlight = rowData.has_webshop_account ? '#E6FFE6' : '';
                row.style.backgroundColor = highlight;
            });
        };
        applyRowColors();  // Initial run after render

        // Add MutationObserver only once
        const scrollArea = document.querySelector('.dt-scrollable');
        if (scrollArea && !scrollArea._observerAttached) {
            const observer = new MutationObserver(() => {
                applyRowColors();  // Reapply styles on changes
            });
            observer.observe(scrollArea, {
                childList: true,
                subtree: true,
            });
            scrollArea._observerAttached = true;
        }
    }
};


function open_new_lead_dialog() {
    const today = frappe.datetime.get_today();
    const d = new frappe.ui.Dialog({
        'title': __("Create a new Lead (Customer, Contact, Address, Contact Note)"),
        'fields': [
            // Section 1: Customer
            { fieldtype: "Section Break" },
            { fieldtype: "Link", fieldname: "existing_customer", label: __("Existing Customer"), options: "Customer",
                get_query: function() {
                    return {
                        'filters': {
                            'disabled': ["!=", 1],
                            'is_internal_customer': 0
                        }
                    };
                },
                change: function() { fetch_customer_fields(d); } },
            { fieldtype: "Link", fieldname: "account_manager", label: __("Sales Manager"), options: "User", reqd: 1 },
            { fieldtype: "Link", fieldname: "default_currency", label: __("Billing Currency"), options: "Currency", reqd: 1 },
            { fieldtype: "Column Break" },
            { fieldtype: "Data", fieldname: "customer_name", label: __("Full Customer Name"), reqd: 1 },
            { fieldtype: "Link", fieldname: "territory", label: __("Territory"), options: "Territory", reqd: 1,
                get_query: function() {
                    return {
                        'filters': {
                            'is_group': 0,
                            'name': ["not in", ["All Territories", "Rest of the World", "Rest of Europe"]]
                        }
                    };
                }
             },
            { fieldtype: "Link", fieldname: "default_price_list", label: __("Default Price List"), options: "Price List", reqd: 1,
                get_query: function() {
                    return {
                        'filters': {
                            'enabled': 1,
                        }
                    };
                }
             },
            { fieldtype: "Column Break" },
            { fieldtype: "Link", fieldname: "company", label: __("Default Company"), options: "Company", reqd: 1 },
            { fieldtype: "Select", fieldname: "language", label: __("Print Language"), options: "en\nde\nfr", reqd: 1, default: "en" },

            // Section 2: Contact
            { fieldtype: "Section Break" },
            { fieldtype: "Data", fieldname: "first_name", label: __("First Name"), reqd: 1 },
            { fieldtype: "Select", fieldname: "salutation", label: __("Salutation (Ms. / Mr.)"), options: "\nMs.\nMr." },
            { fieldtype: "Column Break" },
            { fieldtype: "Data", fieldname: "last_name", label: __("Last Name") },
            { fieldtype: "Select", fieldname: "title", label: __("Title (Dr. / Prof. / Prof. Dr.)"), options: "\nDr.\nProf.\nProf. Dr." },
            { fieldtype: "Column Break" },
            { fieldtype: "Data", fieldname: "email", label: __("Email Address") },
            { fieldtype: "Data", fieldname: "phone", label: __("Phone") },

            // Section 3: Address
            { fieldtype: "Section Break" },
            { fieldtype: "Data", fieldname: "address_line1", label: __("Address Line 1"), reqd: 1 },
            { fieldtype: "Data", fieldname: "address_line2", label: __("Address Line 2") },
            { fieldtype: "Column Break" },
            { fieldtype: "Data", fieldname: "postal_code", label: __("Postal Code"), reqd: 1 },
            { fieldtype: "Data", fieldname: "city", label: __("City/Town"), reqd: 1 },
            { fieldtype: "Column Break" },
            { fieldtype: "Data", fieldname: "state", label: __("State") },
            { fieldtype: "Link", fieldname: "country", label: __("Country"), options: "Country", reqd: 1 },

            // Section 4: Contact Note
            { fieldtype: "Section Break" },
            { fieldtype: "Date", fieldname: "note_date", label: __("Date"), default: today },
            { fieldtype: "Column Break" },
            { fieldtype: "Select", fieldname: "contact_note_type", label: __("Contact Note Type"), options: "\nEmail\nVisit\nPhone\nVideo Call\nConference\nMarketing\nOther" },
            { fieldtype: "Column Break" },
            { fieldtype: "Small Text", fieldname: "note", label: __("Note") },
        ],
        'primary_action_label': __("Create"),
        primary_action(values) {
            // Validate Contact Note fields
            if (values.note && (!values.note_date || !values.contact_note_type)) {
                frappe.msgprint(__("If Note is filled, Date and Contact Note Type are required."));
                return;
            }
            frappe.call({
                'method': "microsynth.microsynth.marketing.create_lead_with_contact_address_note",
                'args': { 'data': values },
                'freeze': true,
                'freeze_message': __("Creating Lead ..."),
                'callback': function(r) {
                    if (r.message && r.message.customer) {
                        frappe.set_route("Form", "Customer", r.message.customer);
                    }
                }
            });
            d.hide();
        }
    });
    d.show();

    // Force wider dialog
    setTimeout(() => {
        let modals = document.getElementsByClassName('modal-dialog');
        if (modals.length > 0) {
            modals[modals.length - 1].style.width = '900px';
        }
    }, 300);
}

function fetch_customer_fields(dialog) {
    const customer = dialog.get_value("existing_customer");
    if (!customer) return;
    frappe.call({
        'method': "frappe.client.get",
        'args': {
            'doctype': "Customer",
            'name': customer
        },
        'callback': function(r) {
            if (r.message) {
                dialog.set_value("customer_name", r.message.customer_name);
                dialog.set_value("company", r.message.default_company);
                dialog.set_value("account_manager", r.message.account_manager);
                dialog.set_value("territory", r.message.territory);
                dialog.set_value("default_currency", r.message.default_currency);
                dialog.set_value("default_price_list", r.message.default_price_list);
                dialog.set_value("language", r.message.language || "en");
            }
        }
    });
}

File: microsynth/microsynth/report/customer_finder/customer_finder.py
# Copyright (c) 2022, Microsynth, libracore and contributors and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
import json


def execute(filters=None):
    columns, data = get_columns(filters), get_data(filters)
    return columns, data


def get_columns(filters):
    return [
        {"label": _("Customer ID"), "fieldname": "customer_id", "fieldtype": "Link", "options": "Customer", "width": 75},
        {"label": _("Customer"), "fieldname": "customer", "fieldtype": "Data", "width": 125},
        {"label": _("Type"), "fieldname": "address_type", "fieldtype": "Data", "width": 35},
        {"label": _("Contact ID"), "fieldname": "contact_id", "fieldtype": "Link", "options":"Contact", "width": 75},
        {"label": _("First Name"), "fieldname": "first_name", "fieldtype": "Data", "width": 75},
        {"label": _("Last Name"), "fieldname": "last_name", "fieldtype": "Data", "width": 75},
        {"label": _("Institute"), "fieldname": "institute", "fieldtype": "Data", "width": 100},
        {"label": _("Institute key"), "fieldname": "institute_key", "fieldtype": "Data", "width": 100},
        {"label": _("Department"), "fieldname": "department", "fieldtype": "Data", "width": 100},
        {"label": _("Group Leader"), "fieldname": "group_leader", "fieldtype": "Data", "width": 75},
        {"label": _("Country"), "fieldname": "country", "fieldtype": "Data", "width": 75},
        {"label": _("City"), "fieldname": "city", "fieldtype": "Data", "width": 75},
        {"label": _("Street"), "fieldname": "street", "fieldtype": "Data", "width": 100},
        {"label": _("Email"), "fieldname": "email_id", "fieldtype": "Data", "width": 120},
        {"label": _("Sales Manager"), "fieldname": "account_manager", "fieldtype": "Data", "width": 70},
        {"label": _("Price List"), "fieldname": "price_list", "fieldtype": "Link", "options":"Price List", "width": 120},
        {"label": _("Tax ID"), "fieldname": "tax_id", "fieldtype": "Data", "width": 110},
        {"label": _("Contact created"), "fieldname": "contact_created", "fieldtype": "Date", "width": 125},
    ]


def get_data(filters):
    
    if type(filters) == str:
        filters = json.loads(filters)
    elif type(filters) == dict:
        pass
    else:
        filters = dict(filters)

    criteria = ""

    hasFilters = False

    if not 'include_disabled' in filters:
        criteria += """ AND `tabCustomer`.`disabled` <> 1 """
    else:
        hasFilters = True

    if 'contact_name' in filters:
        criteria += """ AND `tabContact`.`name` LIKE '%{0}%' """.format(filters['contact_name'])
        hasFilters = True

    if 'contact_full_name' in filters:
        criteria += """ AND `tabContact`.`full_name` LIKE '%{0}%' """.format(filters['contact_full_name'])
        hasFilters = True

    if 'contact_email' in filters:
        criteria += """ AND `tabContact`.`email_id` LIKE '%{0}%' """.format(filters['contact_email'])
        hasFilters = True

    if 'customer' in filters:
        criteria += """ AND `tabCustomer`.`customer_name` LIKE '%{0}%' """.format(filters['customer'])
        hasFilters = True
    
    if 'customer_id' in filters:
        criteria += """ AND `tabCustomer`.`name` = '{0}' """.format(filters['customer_id'])
        hasFilters = True

    if 'contact_institute' in filters:
        criteria += """ AND `tabContact`.`institute` LIKE '%{0}%' """.format(filters['contact_institute'])
        hasFilters = True

    if 'contact_institute_key' in filters:
        criteria += """ AND `tabContact`.`institute_key` LIKE '%{0}%' """.format(filters['contact_institute_key'])
        hasFilters = True

    if 'contact_department' in filters:
        criteria += """ AND `tabContact`.`department` LIKE '%{0}%' """.format(filters['contact_department'])
        hasFilters = True

    if 'contact_group_leader' in filters:
        criteria += """ AND `tabContact`.`group_leader` LIKE '%{0}%' """.format(filters['contact_group_leader'])
        hasFilters = True

    if 'address_country' in filters:
        criteria += """ AND `tabAddress`.`country` LIKE '%{0}%' """.format(filters['address_country'])
        hasFilters = True

    if 'address_city' in filters:
        criteria += """ AND `tabAddress`.`city` LIKE '%{0}%' """.format(filters['address_city'])
        hasFilters = True

    if 'address_street' in filters:
        criteria += """ AND `tabAddress`.`address_line1` LIKE '%{0}%' """.format(filters['address_street'])
        hasFilters = True

    if 'price_list' in filters:
        criteria += """ AND `tabCustomer`.`default_price_list` LIKE '%{0}%' """.format(filters['price_list'])
        hasFilters = True

    if 'tax_id' in filters:
        criteria += """ AND `tabCustomer`.`tax_id` LIKE '%{0}%' """.format(filters['tax_id'])
        hasFilters = True

    if 'account_manager' in filters:
        criteria += """ AND `tabCustomer`.`account_manager` LIKE '%{0}%' """.format(filters['account_manager'])
        hasFilters = True

    if 'contact_status' in filters and filters['contact_status']:
        criteria += """ AND `tabContact`.`status` = '{0}' """.format(filters['contact_status'])
        hasFilters = True

    data = []

    if hasFilters:
        sql_query = """SELECT
            `tabCustomer`.`name` AS `customer_id`,
            `tabCustomer`.`customer_name` AS `customer`,
            `tabAddress`.`address_type` AS `address_type`,
            `tabContact`.`name` AS `contact_id`,
            `tabContact`.`first_name` AS `first_name`,
            `tabContact`.`last_name` AS `last_name`,
            `tabContact`.`email_id` AS `email_id`,
            `tabContact`.`institute` AS `institute`,
            `tabContact`.`department` AS `department`,
            `tabContact`.`group_leader` AS `group_leader`,
            `tabContact`.`institute_key` AS `institute_key`,
            `tabAddress`.`address_line1` AS `address_line1`,
            `tabAddress`.`city` AS `city`,
            `tabAddress`.`country` AS `country`,
            `tabCustomer`.`account_manager` AS `account_manager`,
            `tabCustomer`.`default_price_list` as `price_list`,
            `tabCustomer`.`tax_id` as `tax_id`,
            `tabContact`.`creation` AS `contact_created`

            FROM `tabContact`
            LEFT JOIN `tabDynamic Link` AS `tDLA` ON `tDLA`.`parent` = `tabContact`.`name` 
                                                AND `tDLA`.`parenttype`  = "Contact" 
                                                AND `tDLA`.`link_doctype` = "Customer"
            LEFT JOIN `tabCustomer` ON `tabCustomer`.`name` = `tDLA`.`link_name` 
            LEFT JOIN `tabAddress` ON `tabContact`.`address` = `tabAddress`.`name`
            
            WHERE TRUE
                {criteria}
        """.format(criteria=criteria)

        fetched_data = frappe.db.sql(sql_query, as_dict = True)

        for d in fetched_data:
            entry = {
                "customer_id": d.customer_id,
                "customer": d.customer,
                "address_type": d.address_type,
                "contact_id": d.contact_id,
                "first_name": d.first_name,
                "last_name": d.last_name,
                "email_id": d.email_id,
                "institute": d.institute,
                "institute_key": d.institute_key,
                "department": d.department,
                "group_leader": d.group_leader,
                "country": d.country,
                "city": d.city,
                "street": d.address_line1,
                "account_manager": d.account_manager,
                "price_list": d.price_list,
                "tax_id": d.tax_id,
                "contact_created": d.contact_created
            }
            data.append(entry)

    return data
